pypsa/optimization/constraints.py

```diff
@@ -286,29 +351,122 @@ def define_operational_constraints_for_committables(
+
+        p_nom_max_vals = c.da.p_nom_max.sel(name=com_ext_i)
+        max_pu_vals = max_pu.sel(name=com_ext_i).max("snapshot")
+
+        big_m_default = options.params.optimize.committable_big_m
+        if big_m_default is None:
+            big_m_default = _infer_big_m_scale(n, component)
+        else:
+            if not np.isfinite(big_m_default):
+                msg = f"options.params.optimize.committable_big_m must be finite, got {big_m_default}."
+                raise ValueError(msg)
+            if big_m_default <= 0:
+                msg = f"options.params.optimize.committable_big_m must be positive, got {big_m_default}."
+                raise ValueError(msg)
+
+        fallback_values = big_m_default * max_pu_vals.fillna(1)
+        M_values = xr.where(
+            isfinite(p_nom_max_vals) & (p_nom_max_vals > 0),
+            p_nom_max_vals * max_pu_vals,
+            fallback_values,
+        )
+        p_ext = p.sel(name=com_ext_i)
+        status_ext = status.sel(name=com_ext_i)
+        p_nom_ext = p_nom_var.sel(name=com_ext_i)
+        min_pu_ext = min_pu.sel(name=com_ext_i)
+        max_pu_ext = max_pu.sel(name=com_ext_i)
+
+        active_ext = active.sel(name=com_ext_i)
+        lhs_lower = (1, p_ext), (-min_pu_ext, p_nom_ext), (-M_values, status_ext)
+        n.model.add_constraints(
+            lhs_lower,
+            ">=",
+            -M_values,
+            name=f"{c.name}-com-ext-p-lower",
+            mask=active_ext,
+        )
 
-    # upper dispatch level limit
-    lhs_tuple = (1, p), (-upper_p, status)
-    n.model.add_constraints(
-        lhs_tuple, "<=", 0, name=f"{c.name}-com-p-upper", mask=active
-    )
+        lhs_upper = (1, p_ext), (-M_values, status_ext)
+        n.model.add_constraints(
+            lhs_upper,
+            "<=",
+            0,
+            name=f"{c.name}-com-ext-p-upper-bigM",
+            mask=active_ext,
+        )
+
+        lhs_upper_cap = (1, p_ext), (-max_pu_ext, p_nom_ext)
+        n.model.add_constraints(
+            lhs_upper_cap,
+            "<=",
+            0,
+            name=f"{c.name}-com-ext-p-upper-cap",
+            mask=active_ext,
+        )
+
+        dims_excl_name = [dim for dim in min_pu_ext.dims if dim != "name"]
+        if dims_excl_name:
+            nonneg_mask = (min_pu_ext >= 0).all(dim=dims_excl_name)
+        else:
+            nonneg_mask = min_pu_ext >= 0
+
+        if nonneg_mask.any().item():
+            nonneg_idx = nonneg_mask.to_series()
+            nonneg_idx = nonneg_idx[nonneg_idx].index
+            if len(nonneg_idx) > 0:
+                p_nonneg = p_ext.sel(name=nonneg_idx)
+                active_nonneg = active_ext.sel(name=nonneg_idx)
+                lhs_nonneg = ((1, p_nonneg),)
+                n.model.add_constraints(
+                    lhs_nonneg,
+                    ">=",
+                    0,
+                    name=f"{c.name}-com-ext-p-lower-nonneg",
+                    mask=active_nonneg,
+                )
```

Comment:
I am not fully clear about the nonneg part. Don't you need a similar triple of constraints for the lower bound that you have for the upper one?